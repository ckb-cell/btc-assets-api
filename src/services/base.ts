import axios, { AxiosInstance } from 'axios';
import { FastifyBaseLogger } from 'fastify';

export abstract class BaseRequestService {
  protected request: AxiosInstance;

  constructor(baseURL: string, headers?: Record<string, string>) {
    this.request = axios.create({
      baseURL,
      headers,
    });
  }

  public setLogger(logger: FastifyBaseLogger) {
    const constructor = this.constructor.name;
    this.request.interceptors.request.use((config) => {
      logger.info(`[${constructor}] ${JSON.stringify(config.data)}`);
      return config;
    });

    this.request.interceptors.response.use(
      (response) => {
        logger.info(`[${constructor}] ${response.status} ${JSON.stringify(response.data)}`);
        return response;
      },
      (error) => {
        logger.error(
          `[${constructor}] ${error.response?.status} ${JSON.stringify(error.response?.data)}`,
        );
        return Promise.reject(error);
      },
    );
  }
}
