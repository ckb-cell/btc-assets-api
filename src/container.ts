import { createContainer, InjectionMode, asValue, asClass, asFunction } from 'awilix';
import { Redis } from 'ioredis';
import pino from 'pino';
import Bitcoind from './services/bitcoind';
import ElectrsAPI from './services/electrs';
import { env } from './env';
import TransactionManager from './services/transaction';
import Paymaster from './services/paymaster';
import { RPC as CkbRPC, Indexer as CkbIndexer } from '@ckb-lumos/lumos';
import Unlocker from './services/unlocker';

export interface Cradle {
  env: typeof env;
  logger: pino.BaseLogger;
  redis: Redis;
  ckbRpc: CkbRPC;
  ckbIndexer: CkbIndexer;
  bitcoind: Bitcoind;
  electrs: ElectrsAPI;
  paymaster: Paymaster;
  unlocker: Unlocker;
  transactionManager: TransactionManager;
}

const container = createContainer<Cradle>({
  injectionMode: InjectionMode.PROXY,
  strict: true,
});

container.register({
  env: asValue(env),
  logger: asValue(pino()),
  redis: asValue(
    new Redis(env.REDIS_URL, {
      maxRetriesPerRequest: null,
    }),
  ),
  ckbRpc: asFunction(() => new CkbRPC(env.CKB_RPC_URL)).singleton(),
  ckbIndexer: asFunction(() => new CkbIndexer(env.CKB_INDEXER_URL)).singleton(),
  bitcoind: asClass(Bitcoind).singleton(),
  electrs: asClass(ElectrsAPI).singleton(),
  paymaster: asClass(Paymaster).singleton(),
  transactionManager: asClass(TransactionManager).singleton(),
  unlocker: asClass(Unlocker).singleton(),
});

export default container;
