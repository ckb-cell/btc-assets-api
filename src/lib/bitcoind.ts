import axios, { AxiosInstance } from 'axios';

export interface DescriptorInfo {
  descriptor: string;
  checksum: string;
  isrange: boolean;
  issolvable: boolean;
  hasprivatekeys: boolean;
}

export interface Unspent {
  txid: string;
  vout: number;
  address: string;
  label: string;
  scriptPubKey: string;
  amount: number;
  confirmations: number;
  redeemScript: string;
  spendable: boolean;
  solvable: boolean;
  safe: boolean;
}

export interface AddressInfo {
  address: string;
  scriptPubKey: string;
  ismine: boolean;
  iswatchonly: boolean;
  solvable: boolean;
  desc: string;
  isscript: boolean;
  iswitness: boolean;
  witness_version: number;
  witness_program: string;
  script: string;
  hex: string;
  pubkeys: string[];
  sigsrequired: number;
  pubkey: string;
  iscompressed: boolean;
  ischange: boolean;
  timestamp: number;
  hdkeypath: string;
  hdmasterkeyid: string;
}

export default class Bitcoind {
  private request: AxiosInstance;

  constructor(baseURL: string, username: string, password: string) {
    const credentials = `${username}:${password}`;
    const token = Buffer.from(credentials, 'utf-8').toString('base64');
    this.request = axios.create({
      baseURL,
      headers: {
        Authorization: `Basic ${token}`,
      },
    });
  }

  private async callMethod<T>(method: string, params: unknown): Promise<T> {
    const response = await this.request.post('', {
      jsonrpc: '1.0',
      id: Date.now(),
      method,
      params,
    });
    return response.data.result;
  }

  // https://developer.bitcoin.org/reference/rpc/getdescriptorinfo.html
  public async getDescriptorInfo(descriptor: string) {
    return this.callMethod<DescriptorInfo>('getdescriptorinfo', [descriptor]);
  }

  // https://developer.bitcoin.org/reference/rpc/importdescriptors.html
  public async importDescriptors(descriptors: DescriptorInfo[]) {
    return this.callMethod<void>('importdescriptors', [
      descriptors.map((d) => ({
        desc: d.descriptor,
        timestamp: 'now',
        watchonly: true,
      })),
    ]);
  }

  // https://developer.bitcoin.org/reference/rpc/getaddressinfo.html
  public async getAddressInfo(address: string) {
    return this.callMethod<AddressInfo>('getaddressinfo', [address]);
  }

  // https://developer.bitcoin.org/reference/rpc/listunspent.html
  public async listUnspent(address: string) {
    return this.callMethod<void>('listunspent', [0, 9999999, [address]]);
  }
}
