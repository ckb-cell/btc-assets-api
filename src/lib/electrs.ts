import axios, { AxiosInstance } from 'axios';

export interface UTXO {
  txid: string;
  vout: number;
  value: number;
  status: {
    confirmed: boolean;
    block_height: number;
    block_hash: string;
    block_time: number;
  };
}

export interface Balance {
  address: string;
  satoshi: number;
  pendingSatoshi: number;
  utxoCount: number;
}

export default class ElectrsAPI {
  private request: AxiosInstance;

  constructor(baseURL: string) {
    this.request = axios.create({
      baseURL,
    });
  }

  async getUtxoByAddress(address: string): Promise<UTXO[]> {
    const response = await this.request.get(`/address/${address}/utxo`);
    return response.data;
  }

  async getBalanceByAddress(address: string): Promise<Balance> {
    const utxos = await this.getUtxoByAddress(address);
    return utxos.reduce(
      (acc, utxo) => {
        if (utxo.status.confirmed) {
          acc.satoshi += utxo.value;
        } else {
          acc.pendingSatoshi += utxo.value;
        }
        return acc;
      },
      {
        address,
        satoshi: 0,
        pendingSatoshi: 0,
        utxoCount: utxos.length,
      },
    );
  }
}
