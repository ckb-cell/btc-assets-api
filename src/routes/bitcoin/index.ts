import { FastifyPluginCallback } from 'fastify';
import { Server } from 'http';
import infoRoute from './info';
import blockRoutes from './block';
import transactionRoutes from './transaction';
import addressRoutes from './address';
import container from '../../container';
import { ZodTypeProvider } from 'fastify-type-provider-zod';
import BitcoinClient from '../../services/bitcoin';
import feesRoutes from './fees';

const bitcoinRoutes: FastifyPluginCallback<Record<never, never>, Server, ZodTypeProvider> = (fastify, _, done) => {
  fastify.decorate('bitcoin', container.resolve<BitcoinClient>('bitcoin'));

  fastify.register(infoRoute);
  fastify.register(blockRoutes, { prefix: '/block' });
  fastify.register(transactionRoutes, { prefix: '/transaction' });
  fastify.register(addressRoutes, { prefix: '/address' });
  fastify.register(feesRoutes, { prefix: '/fees' });
  done();
};

export default bitcoinRoutes;
