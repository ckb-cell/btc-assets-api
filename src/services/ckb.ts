import { Collector, getSporeTypeScript, getUniqueTypeScript, getXudtTypeScript, sendCkbTx } from '@rgbpp-sdk/ckb';
import { Cradle } from '../container';
import { Indexer, RPC, Script } from '@ckb-lumos/lumos';
import { z } from 'zod';
import * as Sentry from '@sentry/node';
import { serializeScript } from '@nervosnetwork/ckb-sdk-utils';
import { computeScriptHash } from '@ckb-lumos/lumos/utils';
import { decodeUniqueCellData } from '../utils/xudt';

// https://github.com/nervosnetwork/ckb/blob/develop/rpc/src/error.rs#L33
export enum CKBRPCErrorCodes {
  /// (-1): CKB internal errors are considered to never happen or only happen when the system
  /// resources are exhausted.
  CKBInternalError = -1,
  /// (-2): The CKB method has been deprecated and disabled.
  ///
  /// Set `rpc.enable_deprecated_rpc` to `true` in the config file to enable all deprecated
  /// methods.
  Deprecated = -2,
  /// (-3): Error code -3 is no longer used.
  ///
  /// Before v0.35.0, CKB returns all RPC errors using the error code -3. CKB no longer uses
  /// -3 since v0.35.0.
  Invalid = -3,
  /// (-4): The RPC method is not enabled.
  ///
  /// CKB groups RPC methods into modules, and a method is enabled only when the module is
  /// explicitly enabled in the config file.
  RPCModuleIsDisabled = -4,
  /// (-5): DAO related errors.
  DaoError = -5,
  /// (-6): Integer operation overflow.
  IntegerOverflow = -6,
  /// (-7): The error is caused by a config file option.
  ///
  /// Users have to edit the config file to fix the error.
  ConfigError = -7,
  /// (-101): The CKB local node failed to broadcast a message to its peers.
  P2PFailedToBroadcast = -101,
  /// (-200): Internal database error.
  ///
  /// The CKB node persists data to the database. This is the error from the underlying database
  /// module.
  DatabaseError = -200,
  /// (-201): The chain index is inconsistent.
  ///
  /// An example of an inconsistent index is that the chain index says a block hash is in the chain
  /// but the block cannot be read from the database.
  ///
  /// This is a fatal error usually due to a serious bug. Please back up the data directory and
  /// re-sync the chain from scratch.
  ChainIndexIsInconsistent = -201,
  /// (-202): The underlying database is corrupt.
  ///
  /// This is a fatal error usually caused by the underlying database used by CKB. Please back up
  /// the data directory and re-sync the chain from scratch.
  DatabaseIsCorrupt = -202,
  /// (-301): Failed to resolve the referenced cells and headers used in the transaction, as inputs or
  /// dependencies.
  TransactionFailedToResolve = -301,
  /// (-302): Failed to verify the transaction.
  TransactionFailedToVerify = -302,
  /// (-1000): Some signatures in the submit alert are invalid.
  AlertFailedToVerifySignatures = -1000,
  /// (-1102): The transaction is rejected by the outputs validator specified by the RPC parameter.
  PoolRejectedTransactionByOutputsValidator = -1102,
  /// (-1103): Pool rejects some transactions which seem contain invalid VM instructions. See the issue
  /// link in the error message for details.
  PoolRejectedTransactionByIllTransactionChecker = -1103,
  /// (-1104): The transaction fee rate must be greater than or equal to the config option `tx_pool.min_fee_rate`
  ///
  /// The fee rate is calculated as:
  ///
  /// ```text
  /// fee / (1000 * tx_serialization_size_in_block_in_bytes)
  /// ```
  PoolRejectedTransactionByMinFeeRate = -1104,
  /// (-1105): The in-pool ancestors count must be less than or equal to the config option `tx_pool.max_ancestors_count`
  ///
  /// Pool rejects a large package of chained transactions to avoid certain kinds of DoS attacks.
  PoolRejectedTransactionByMaxAncestorsCountLimit = -1105,
  /// (-1106): The transaction is rejected because the pool has reached its limit.
  PoolIsFull = -1106,
  /// (-1107): The transaction is already in the pool.
  PoolRejectedDuplicatedTransaction = -1107,
  /// (-1108): The transaction is rejected because it does not make sense in the context.
  ///
  /// For example, a cellbase transaction is not allowed in `send_transaction` RPC.
  PoolRejectedMalformedTransaction = -1108,
  /// (-1109): The transaction is expired from tx-pool after `expiry_hours`.
  TransactionExpired = -1109,
  /// (-1110): The transaction exceeded maximum size limit.
  PoolRejectedTransactionBySizeLimit = -1110,
  /// (-1111): The transaction is rejected for RBF checking.
  PoolRejectedRBF = -1111,
  /// (-1112): The transaction is rejected for ref cell consuming.
  PoolRejectedInvalidated = -1112,
  /// (-1200): The indexer error.
  Indexer = -1200,
}

export class CKBRpcError extends Error {
  private messageSchema = z.object({
    code: z.number(),
    message: z.string(),
  });

  public code?: number;
  public message: string;

  constructor(message: string) {
    super(message);
    this.name = 'CKBRpcError';
    this.message = message;

    try {
      const error = JSON.parse(message);
      const parsed = this.messageSchema.safeParse(error);
      if (parsed.success) {
        this.code = parsed.data.code;
        this.message = parsed.data.message;
      }
    } catch (e) {
      Sentry.captureException(e);
    }
  }
}

export default class CKBClient {
  public rpc: RPC;
  public indexer: Indexer;

  constructor(private cradle: Cradle) {
    this.rpc = new RPC(cradle.env.CKB_RPC_URL);
    this.indexer = new Indexer(cradle.env.CKB_RPC_URL);
  }

  private async getAllUniqueCellTxs() {
    console.log('Fetching all unique cell transactions');
    const scripts = this.getScripts();
    const result = await this.rpc.getTransactions(
      {
        script: {
          codeHash: scripts.UNIQUE.codeHash,
          hashType: scripts.UNIQUE.hashType,
          args: '0x',
        },
        scriptType: 'type',
      },
      'desc',
      '0xffff', // 0xffff basically means no limit
    );
    // get all transactions that have the xudt type cell and unique cell
    const batchRequest = this.rpc.createBatchRequest(
      result.objects.filter((tx) => tx.ioType === 'output').map((tx) => ['getTransaction', tx.txHash]),
    );
    const txs = (await batchRequest.exec()) as CKBComponents.TransactionWithStatus[];
    return txs;
  }

  /**
   * Get the ckb script configs
   */
  public getScripts() {
    const isMainnet = this.cradle.env.NETWORK === 'mainnet';
    const xudtTypeScript = getXudtTypeScript(isMainnet);
    const sporeTypeScript = getSporeTypeScript(isMainnet);
    const uniqueCellTypeScript = getUniqueTypeScript(isMainnet);
    return {
      XUDT: xudtTypeScript,
      SPORE: sporeTypeScript,
      UNIQUE: uniqueCellTypeScript,
    };
  }

  /**
   * Get the unique cell of the given xudt type
   * @param script - the xudt type script
   */
  public async getUniqueCellByType(script: Script) {
    const scripts = this.getScripts();
    const txs = await this.getAllUniqueCellTxs();
    for (const tx of txs) {
      const xudtCellIndex = tx.transaction.outputs.findIndex(
        (cell) => cell.type && serializeScript(cell.type) === serializeScript(script),
      );
      const uniqueCellIndex = tx.transaction.outputs.findIndex(
        (cell) =>
          cell.type &&
          serializeScript({
            ...cell.type,
            args: '',
          }) === serializeScript(scripts.UNIQUE),
      );
      if (xudtCellIndex !== -1 && uniqueCellIndex !== -1) {
        const encodeData = tx.transaction.outputsData[uniqueCellIndex];
        const typeHash = computeScriptHash(tx.transaction.outputs[xudtCellIndex].type!);
        const data = decodeUniqueCellData(encodeData);
        return {
          ...data,
          typeHash,
        };
      }
    }
    return null;
  }

  /**
   * Wait for the ckb transaction to be confirmed
   * @param txHash - the ckb transaction hash
   */
  public waitForTranscationConfirmed(txHash: string) {
    // eslint-disable-next-line no-async-promise-executor
    return new Promise(async (resolve) => {
      try {
        const transaction = await this.rpc.getTransaction(txHash);
        const { status } = transaction.txStatus;
        if (status === 'committed') {
          resolve(txHash);
        } else {
          setTimeout(() => {
            resolve(this.waitForTranscationConfirmed(txHash));
          }, 1000);
        }
      } catch (e) {
        Sentry.withScope((scope) => {
          scope.setTag('ckb_txhash', txHash);
          scope.captureException(e);
        });
        setTimeout(() => {
          resolve(this.waitForTranscationConfirmed(txHash));
        }, 1000);
      }
    });
  }

  /**
   * Send a ckb transaction
   * @param signedTx - the signed ckb transaction
   */
  public async sendTransaction(signedTx: CKBComponents.RawTransaction): Promise<string> {
    try {
      const txHash = await sendCkbTx({
        collector: new Collector({
          ckbNodeUrl: this.cradle.env.CKB_RPC_URL,
          ckbIndexerUrl: this.cradle.env.CKB_RPC_URL,
        }),
        signedTx,
      });
      return txHash;
    } catch (err) {
      if (err instanceof Error) {
        const rpcError = new CKBRpcError(err.message);
        rpcError.stack = err.stack;
        throw rpcError;
      }
      throw err;
    }
  }
}
