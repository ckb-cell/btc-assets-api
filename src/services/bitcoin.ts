import mempoolJS from '@mempool/mempool.js';
import { Cradle } from '../container';
import { Block, ChainInfo, Transaction, UTXO } from '../routes/bitcoin/types';
import { NetworkType } from '../constants';
import { AxiosError } from 'axios';
import Electrs from '../utils/electrs';
import * as Sentry from '@sentry/node';

// https://github.com/mempool/electrs/blob/d4f788fc3d7a2b4eca4c5629270e46baba7d0f19/src/errors.rs#L6
export enum MempoolErrorMessage {
  Connection = 'Connection error',
  Interrupt = 'Interruption by external signal',
  TooManyUtxos = 'Too many unspent transaction outputs',
  TooManyTxs = 'Too many history transactions',
  ElectrumClient = 'Electrum client error',
}

export enum BitcoinClientErrorCode {
  Connection = 0x1000, // 4096
  Interrupt = 0x1001, // 4097
  TooManyUtxos = 0x1002, // 4098
  TooManyTxs = 0x1003, // 4099
  ElectrumClient = 0x1004, // 4100

  MempoolUnknown = 0x1111, // 4369
}

const BitcoinClientErrorMap = {
  [MempoolErrorMessage.Connection]: BitcoinClientErrorCode.Connection,
  [MempoolErrorMessage.Interrupt]: BitcoinClientErrorCode.Interrupt,
  [MempoolErrorMessage.TooManyUtxos]: BitcoinClientErrorCode.TooManyUtxos,
  [MempoolErrorMessage.TooManyTxs]: BitcoinClientErrorCode.TooManyTxs,
  [MempoolErrorMessage.ElectrumClient]: BitcoinClientErrorCode.ElectrumClient,
};

export class BitcoinClientAPIError extends Error {
  public statusCode = 500;
  public errorCode: BitcoinClientErrorCode;

  constructor(message: string) {
    super(message);
    this.name = this.constructor.name;

    const errorKey = Object.keys(BitcoinClientErrorMap).find((msg) => message.startsWith(msg));
    this.errorCode = BitcoinClientErrorMap[errorKey as MempoolErrorMessage] ?? BitcoinClientErrorCode.MempoolUnknown;
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const wrapTry = async <T extends (...args: any) => Promise<any>>(fn: T): Promise<ReturnType<T>> => {
  if (typeof fn !== 'function') {
    throw new Error('wrapTry: fn must be a function');
  }

  try {
    const ret = await fn();
    return ret;
  } catch (err) {
    if ((err as AxiosError).isAxiosError) {
      const error = new BitcoinClientAPIError((err as AxiosError).message);
      if ((err as AxiosError).response) {
        error.statusCode = (err as AxiosError).response?.status || 500;
      }
      throw error;
    }
    throw err;
  }
};

export default class BitcoinClient {
  private cradle: Cradle;
  private mempool: ReturnType<typeof mempoolJS>;
  private electrs?: Electrs;

  constructor(cradle: Cradle) {
    this.cradle = cradle;

    const url = new URL(cradle.env.BITCOIN_MEMPOOL_SPACE_API_URL);
    this.mempool = mempoolJS({
      hostname: url.hostname,
      network: cradle.env.NETWORK,
    });

    if (cradle.env.BITCOIN_ELECTRS_API_URL) {
      this.electrs = new Electrs(cradle);
    }
  }

  public async checkNetwork(network: NetworkType) {
    const hash = await this.getBlockHashByHeight(0);
    switch (network) {
      case NetworkType.mainnet:
        // Bitcoin mainnet genesis block hash
        if (hash !== '000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f') {
          throw new Error('Mempool API is not running on mainnet');
        }
        break;
      case NetworkType.testnet:
        // Bitcoin testnet genesis block hash
        if (hash !== '000000000933ea01ad0ee984209779baaec3ced90fa3f408719526f8d77f4943') {
          throw new Error('Mempool API is not running on testnet');
        }
        break;
      default:
    }
  }

  public async getBlockchainInfo(): Promise<ChainInfo> {
    const hash = await this.mempool.bitcoin.blocks.getBlocksTipHash();
    const tip = await this.mempool.bitcoin.blocks.getBlock({ hash });

    const { difficulty, mediantime } = tip;
    return {
      chain: this.cradle.env.NETWORK === 'mainnet' ? 'main' : 'test',
      blocks: tip.height,
      bestblockhash: hash,
      difficulty,
      mediantime,
    };
  }

  public async sendRawTransaction(txhex: string): Promise<string> {
    return wrapTry(async () => {
      try {
        const txid = await this.mempool.bitcoin.transactions.postTx({ txhex });
        return txid as string;
      } catch (err) {
        this.cradle.logger.error(err);
        Sentry.captureException(err);
        if (this.electrs) {
          return this.electrs.sendRawTransaction(txhex);
        }
        throw err;
      }
    });
  }

  public async getUtxoByAddress(address: string): Promise<UTXO[]> {
    return wrapTry(async () => {
      try {
        const utxo = await this.mempool.bitcoin.addresses.getAddressTxsUtxo({ address });
        return utxo.map((utxo) => UTXO.parse(utxo));
      } catch (err) {
        this.cradle.logger.error(err);
        Sentry.captureException(err);
        if (this.electrs) {
          return this.electrs.getUtxoByAddress(address);
        }
        throw err;
      }
    });
  }

  public async getTransactionsByAddress(address: string, after_txid?: string): Promise<Transaction[]> {
    return wrapTry(async () => {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-expect-error after_txid is not defined in the type definition
      const txs = await this.mempool.bitcoin.addresses.getAddressTxs({ address, after_txid });
      return txs.map((tx) => Transaction.parse(tx));
    });
  }

  public async getTransaction(txid: string): Promise<Transaction> {
    return wrapTry(async () => {
      const tx = await this.mempool.bitcoin.transactions.getTx({ txid });
      return Transaction.parse(tx);
    });
  }

  public async getTransactionHex(txid: string): Promise<string> {
    return wrapTry(() => this.mempool.bitcoin.transactions.getTxHex({ txid }));
  }

  public async getBlockByHash(hash: string): Promise<Block> {
    return wrapTry(async () => {
      const block = await this.mempool.bitcoin.blocks.getBlock({ hash });
      return block;
    });
  }

  public async getBlockByHeight(height: number): Promise<Block> {
    return wrapTry(async () => {
      try {
        const hash = await this.mempool.bitcoin.blocks.getBlockHeight({ height });
        const block = await this.mempool.bitcoin.blocks.getBlock({ hash });
        return Block.parse(block);
      } catch (err) {
        this.cradle.logger.error(err);
        Sentry.captureException(err);
        if (this.electrs) {
          const hash = await this.electrs.getBlockHashByHeight(height);
          const block = await this.electrs.getBlockByHash(hash);
          return Block.parse(block);
        }
        throw err;
      }
    });
  }

  public async getBlockHashByHeight(height: number): Promise<string> {
    return wrapTry(async () => {
      try {
        const hash = await this.mempool.bitcoin.blocks.getBlockHeight({ height });
        return hash;
      } catch (err) {
        this.cradle.logger.error(err);
        Sentry.captureException(err);
        if (this.electrs) {
          const hash = await this.electrs.getBlockHashByHeight(height);
          return hash;
        }
        throw err;
      }
    });
  }

  public async getBlockHeaderByHash(hash: string) {
    return wrapTry(async () => {
      try {
        return this.mempool.bitcoin.blocks.getBlockHeader({ hash });
      } catch (err) {
        this.cradle.logger.error(err);
        Sentry.captureException(err);
        if (this.electrs) {
          return this.electrs.getBlockHeaderByHash(hash);
        }
        throw err;
      }
    });
  }

  public async getBlockTxIdsByHash(hash: string): Promise<string[]> {
    return wrapTry(async () => {
      try {
        const txids = await this.mempool.bitcoin.blocks.getBlockTxids({ hash });
        return txids;
      } catch (err) {
        this.cradle.logger.error(err);
        Sentry.captureException(err);
        if (this.electrs) {
          return this.electrs.getBlockTxIdsByHash(hash);
        }
        throw err;
      }
    });
  }

  public async getTip(): Promise<number> {
    return wrapTry(async () => {
      try {
        return this.mempool.bitcoin.blocks.getBlocksTipHeight();
      } catch (err) {
        this.cradle.logger.error(err);
        Sentry.captureException(err);
        if (this.electrs) {
          return this.electrs.getTip();
        }
        throw err;
      }
    });
  }
}
