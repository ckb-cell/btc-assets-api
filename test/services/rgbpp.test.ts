import container from '../../src/container';
import { describe, test, beforeEach, afterEach, vi, expect } from 'vitest';
import RgbppCollector from '../../src/services/rgbpp';
import { UTXO } from '../../src/services/bitcoin/schema';
import { Script } from '@ckb-lumos/base';
import { IndexerCell } from '@ckb-lumos/ckb-indexer/lib/type';
import { buildRgbppLockArgs, genRgbppLockScript } from '@rgbpp-sdk/ckb';

const xudtTypeScript = {
  codeHash: '0x25c29dc317811a6f6f3985a7a9ebc4838bd388d19d0feeecf0bcd60f6c0975bb',
  hashType: 'type',
  args: '0x',
};

describe('RgbppCollector', () => {
  let rgbppCollector: RgbppCollector;

  beforeEach(async () => {
    const cradle = container.cradle;
    rgbppCollector = new RgbppCollector(cradle);
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  test('getRgbppCellsBatchRequest: should return the batch request for the utxos', () => {
    const utxos: UTXO[] = [
      {
        txid: '0x2c4fa9f077b6bf28ef938414f11609a0620b92f23bf7e6bdcb25cc69d9c4b109',
        vout: 0,
        value: 100000000,
        status: { confirmed: true },
      },
      {
        txid: '0x2c4fa9f077b6bf28ef938414f11609a0620b92f23bf7e6bdcb25cc69d9c4b109',
        vout: 1,
        value: 200000000,
        status: { confirmed: true },
      },
    ];

    const batchRequest = rgbppCollector['getRgbppCellsBatchRequest'](utxos);
    expect(batchRequest).toMatchSnapshot();
  });

  test('getRgbppCellsBatchRequest: should be filtered by typeScript', () => {
    const utxos: UTXO[] = [
      {
        txid: '0x2c4fa9f077b6bf28ef938414f11609a0620b92f23bf7e6bdcb25cc69d9c4b109',
        vout: 0,
        value: 100000000,
        status: { confirmed: true },
      },
      {
        txid: '0x2c4fa9f077b6bf28ef938414f11609a0620b92f23bf7e6bdcb25cc69d9c4b109',
        vout: 1,
        value: 200000000,
        status: { confirmed: true },
      },
    ];

    const typeScript: Script = {
      codeHash: '0x25c29dc317811a6f6f3985a7a9ebc4838bd388d19d0feeecf0bcd60f6c0975bb',
      hashType: 'data',
      args: '0x',
    };

    const batchRequest = rgbppCollector['getRgbppCellsBatchRequest'](utxos, typeScript);
    expect(batchRequest).toMatchSnapshot();
  });

  test('collectRgbppUtxoCellsPairs: should return the utxo and the cells', async () => {
    const utxos: UTXO[] = [
      {
        txid: '0x2c4fa9f077b6bf28ef938414f11609a0620b92f23bf7e6bdcb25cc69d9c4b109',
        vout: 0,
        value: 100000000,
        status: { confirmed: true },
      },
      {
        txid: '0x2c4fa9f077b6bf28ef938414f11609a0620b92f23bf7e6bdcb25cc69d9c4b109',
        vout: 1,
        value: 200000000,
        status: { confirmed: true },
      },
    ];
    vi.spyOn(rgbppCollector, 'getRgbppCellsBatchRequest').mockReturnValue({
      exec: vi.fn().mockResolvedValue([
        {
          objects: [
            {
              blockNumber: '0x123',
              outPoint: {
                txHash: '0x',
                index: '0x0',
              },
              output: {
                lock: genRgbppLockScript(
                  buildRgbppLockArgs(0, '0x2c4fa9f077b6bf28ef938414f11609a0620b92f23bf7e6bdcb25cc69d9c4b109'),
                  false,
                ),
                type: xudtTypeScript,
                capacity: '0x123',
              },
              outputData: '0x',
              txIndex: '0x0',
            },
          ] as IndexerCell[],
        },
        {
          objects: [
            {
              blockNumber: '0x456',
              outPoint: {
                txHash: '0x',
                index: '0x0',
              },
              output: {
                lock: genRgbppLockScript(
                  buildRgbppLockArgs(1, '0x2c4fa9f077b6bf28ef938414f11609a0620b92f23bf7e6bdcb25cc69d9c4b109'),
                  false,
                ),
                type: xudtTypeScript,
                capacity: '0x123',
              },
              outputData: '0x',
              txIndex: '0x0',
            },
          ],
        },
      ]),
    });

    const rgbppUtxoCellsPairs = await rgbppCollector.collectRgbppUtxoCellsPairs(utxos);
    expect(rgbppUtxoCellsPairs).toMatchSnapshot();
  });
});
