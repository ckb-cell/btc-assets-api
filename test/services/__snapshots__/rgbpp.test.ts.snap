// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RgbppCollector > collectRgbppUtxoCellsPairs: should return the utxo and the cells 1`] = `
[
  {
    "cells": [
      {
        "blockNumber": "0x123",
        "cellOutput": {
          "capacity": "0x123",
          "lock": {
            "args": "0x0000000009b1c4d969cc25cbbde6f73bf2920b62a00916f1148493ef28bfb677f0a94f2c",
            "codeHash": "0x61ca7a4796a4eb19ca4f0d065cb9b10ddcf002f10f7cbb810c706cb6bb5c3248",
            "hashType": "type",
          },
          "type": {
            "args": "0x",
            "codeHash": "0x25c29dc317811a6f6f3985a7a9ebc4838bd388d19d0feeecf0bcd60f6c0975bb",
            "hashType": "type",
          },
        },
        "data": "0x",
        "outPoint": {
          "index": "0x0",
          "txHash": "0x",
        },
        "txIndex": "0x0",
      },
    ],
    "utxo": {
      "status": {
        "confirmed": true,
      },
      "txid": "0x2c4fa9f077b6bf28ef938414f11609a0620b92f23bf7e6bdcb25cc69d9c4b109",
      "value": 100000000,
      "vout": 0,
    },
  },
  {
    "cells": [
      {
        "blockNumber": "0x456",
        "cellOutput": {
          "capacity": "0x123",
          "lock": {
            "args": "0x0100000009b1c4d969cc25cbbde6f73bf2920b62a00916f1148493ef28bfb677f0a94f2c",
            "codeHash": "0x61ca7a4796a4eb19ca4f0d065cb9b10ddcf002f10f7cbb810c706cb6bb5c3248",
            "hashType": "type",
          },
          "type": {
            "args": "0x",
            "codeHash": "0x25c29dc317811a6f6f3985a7a9ebc4838bd388d19d0feeecf0bcd60f6c0975bb",
            "hashType": "type",
          },
        },
        "data": "0x",
        "outPoint": {
          "index": "0x0",
          "txHash": "0x",
        },
        "txIndex": "0x0",
      },
    ],
    "utxo": {
      "status": {
        "confirmed": true,
      },
      "txid": "0x2c4fa9f077b6bf28ef938414f11609a0620b92f23bf7e6bdcb25cc69d9c4b109",
      "value": 200000000,
      "vout": 1,
    },
  },
]
`;

exports[`RgbppCollector > getRgbppCellsBatchRequest: should be filtered by typeScript 1`] = `
[
  [
    "getCells",
    {
      "filter": {
        "script": {
          "args": "0x",
          "codeHash": "0x25c29dc317811a6f6f3985a7a9ebc4838bd388d19d0feeecf0bcd60f6c0975bb",
          "hashType": "data",
        },
      },
      "script": {
        "args": "0x0000000009b1c4d969cc25cbbde6f73bf2920b62a00916f1148493ef28bfb677f0a94f2c",
        "codeHash": "0x61ca7a4796a4eb19ca4f0d065cb9b10ddcf002f10f7cbb810c706cb6bb5c3248",
        "hashType": "type",
      },
      "scriptType": "lock",
    },
    "desc",
    "0x64",
  ],
  [
    "getCells",
    {
      "filter": {
        "script": {
          "args": "0x",
          "codeHash": "0x25c29dc317811a6f6f3985a7a9ebc4838bd388d19d0feeecf0bcd60f6c0975bb",
          "hashType": "data",
        },
      },
      "script": {
        "args": "0x0100000009b1c4d969cc25cbbde6f73bf2920b62a00916f1148493ef28bfb677f0a94f2c",
        "codeHash": "0x61ca7a4796a4eb19ca4f0d065cb9b10ddcf002f10f7cbb810c706cb6bb5c3248",
        "hashType": "type",
      },
      "scriptType": "lock",
    },
    "desc",
    "0x64",
  ],
]
`;

exports[`RgbppCollector > getRgbppCellsBatchRequest: should return the batch request for the utxos 1`] = `
[
  [
    "getCells",
    {
      "script": {
        "args": "0x0000000009b1c4d969cc25cbbde6f73bf2920b62a00916f1148493ef28bfb677f0a94f2c",
        "codeHash": "0x61ca7a4796a4eb19ca4f0d065cb9b10ddcf002f10f7cbb810c706cb6bb5c3248",
        "hashType": "type",
      },
      "scriptType": "lock",
    },
    "desc",
    "0x64",
  ],
  [
    "getCells",
    {
      "script": {
        "args": "0x0100000009b1c4d969cc25cbbde6f73bf2920b62a00916f1148493ef28bfb677f0a94f2c",
        "codeHash": "0x61ca7a4796a4eb19ca4f0d065cb9b10ddcf002f10f7cbb810c706cb6bb5c3248",
        "hashType": "type",
      },
      "scriptType": "lock",
    },
    "desc",
    "0x64",
  ],
]
`;
